sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,152 @@
+    val st = getOutputCollation(exprs)
+
+    exprs.map(e => castStringType(e, st).getOrElse(e))
+  }
+
+  /**
+   * Based on the data types of the input expressions this method determines
+   * a collation type which the output will have. This function accepts Seq of
+   * any expressions, but will only be affected by collated StringTypes or
+   * complex DataTypes with collated StringTypes (e.g. ArrayType)
+   */
+  def getOutputCollation(expr: Seq[Expression]): StringType = {
+    val explicitTypes = expr.filter(_.isInstanceOf[Collate])
+      .map(_.dataType.asInstanceOf[StringType].collationId)
+      .distinct
+
+    explicitTypes.size match {
+      // We have 1 explicit collation
+      case 1 => StringType(explicitTypes.head)
+      // Multiple explicit collations occurred
+      case size if size > 1 =>
+        throw QueryCompilationErrors
+          .explicitCollationMismatchError(
+            explicitTypes.map(t => StringType(t).typeName)
+          )
+      // Only implicit or default collations present
+      case 0 =>
+        val implicitTypes = expr.map(_.dataType)
+          .filter(hasStringType)
+          .map(extractStringType)
+          .filter(dt => dt.collationId != SQLConf.get.defaultStringType.collationId)
+
+        if (hasMultipleImplicits(implicitTypes)) {
+          throw QueryCompilationErrors.implicitCollationMismatchError()
+        }
+        else {
+          implicitTypes.find(dt => !(dt == SQLConf.get.defaultStringType))
+            .getOrElse(SQLConf.get.defaultStringType)
+        }
+    }
+  }
+
+  /**
+   * This check is always preformed when we have no explicit collation. It returns true
+   * if there are more than one implicit collations. Collations are distinguished by their
+   * collationId.
+   * @param dataTypes
+   * @return
+   */
+  private def hasMultipleImplicits(dataTypes: Seq[StringType]): Boolean =
+    dataTypes.map(_.collationId)
+      .filter(dt => !(dt == SQLConf.get.defaultStringType.collationId)).distinct.size > 1
+
+}
+
+/**
+ * This rule is used to collate all existing expressions related to StringType into a single
+ * collation. Arrays are handled using their elementType and should be cast for these expressions.
+ */
+object PreCollationTypeCasts extends CollationTypeCasts {
+  override val transform: PartialFunction[Expression, Expression] = {
+    case e if !e.childrenResolved => e
+    case sc@(_: In
+             | _: InSubquery
+             | _: CreateArray
+             | _: ComplexTypeMergingExpression
+             | _: ArrayJoin
+             | _: BinaryExpression
+             | _: ConcatWs
+             | _: Substring) =>
+      val newChildren = collateToSingleType(sc.children)
+      sc.withNewChildren(newChildren)
+  }
+}
+
+/**
+ *  This rule is used for managing expressions that have possible implicit casts from different
+ *  types in ImplicitTypeCasts rule.
+ */
+object PostCollationTypeCasts extends CollationTypeCasts {
```

Comment:
why does this rule need two phases?

Reply:
Catalyst rules are executed with a loop, it's OK to wait for the next iteration and adjust the implicit casts.

Reply:
Some expressions might be cast from different types, e.g. (NullType or IntegerType) to StringType, and we want to make sure these are picked up and accordingly cast to the proper collation for expressions that lie into that category (i.e. expressions that extend ImpicitInputTypeCast and ExpectsInputType, as well as BinaryExpressions). It is either this or fine-picking these expressions in ImplicitTypeCasts, but this approach seemed more appropriate. What do you think? (We only do this for ArrayJoin, Substring and BinaryExpressions, as these are the only supported ones for collations so far)

Reply:
Can't `PreCollationTypeCasts` alone do the work?